### PR TITLE
trie parts

### PR DESCRIPTION
**Pull request by mikhailOK**

### CODE REVIEW

File: core/store/src/trie/state_parts.rs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        iterator.seek_nibble_slice(NibbleSlice::from_encoded(&path_begin[..]).0)?;
+        loop {
+            match iterator.next() {
+                None => break,
+                Some(Err(e)) => {
+                    return Err(e);
+                }
+                Some(Ok(_)) => {}
+            }
+            // TODO this is bad for large keys
```
Comment (ilblackdragon) [inline core/store/src/trie/state_parts.rs]: let's add an issue to track?

File: core/store/src/trie/state_parts.rs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    fn visit_nodes_for_size_range(
+        &self,
+        root_hash: &CryptoHash,
+        size_start: u64,
+        size_end: u64,
+    ) -> Result<(), StorageError> {
+        let root_node = self.retrieve_node(&root_hash)?;
+        let path_begin = self.find_path(&root_node, size_start)?;
+        let mut path_end = self.find_path(&root_node, size_end)?;
+        path_end.push(0x10);
```
Comment (Kouprin) [inline core/store/src/trie/state_parts.rs]: `0x10`?

Comment (Kouprin) [reply core/store/src/trie/state_parts.rs]: `PATH_END` constant?

File: core/store/src/trie/state_parts.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+impl Trie {
+    /// Computes the set of trie nodes for a state part.
+    ///
+    /// # Panics
+    /// storage must be a TrieCachingStorage
+    /// part_id must be in [0..num_parts)
+    ///
+    /// # Errors
+    /// StorageError if the storage is corrupted
+    pub fn get_state_part(
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply core/store/src/trie/state_parts.rs]: `get_trie_nodes_for_part`?

**Comment by Kouprin** [comment]
> shouldn't be a problem to remove num_parts from StateRoot
Not too easy. We have to decide how many parts we want to split the state. To do this, we need to know the size of the state + prove it.
Currently this problem solved by adding `num_parts` to StateRoot but I'm sure it's not the best solution.

File: core/store/src/trie/state_parts.rs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+                    .collect::<Vec<_>>();
+
+                for part in parts {
+                    for node in part {
+                        nodes.insert(hash(&node), node);
+                    }
+                }
+                let all_nodes = nodes.into_iter().map(|(_hash, node)| node).collect::<Vec<_>>();
+                let size_of_all = all_nodes.iter().map(|node| node.len()).sum::<usize>();
+                Trie::validate_state_part(&StateRoot { hash: root, num_parts: 1 }, 0, &all_nodes)
```
[the remark this answers is not in the document]
Comment (Kouprin) [reply core/store/src/trie/state_parts.rs]: Would you add this logic to the current PR?

**Pull request merged**